File: backend/angle_utils.py
"""
Angle calculation utilities for pose analysis.
Implements LearnOpenCV AI Fitness Trainer angle calculations.
"""
import numpy as np
from typing import Tuple, List


def angle_at_point(a: Tuple[float, float], b: Tuple[float, float], c: Tuple[float, float]) -> float:
    """
    Calculate angle at point B formed by points A-B-C.
    Used for joint angles like hip-knee-ankle.
    
    Args:
        a: First point (x, y)
        b: Vertex point (x, y) - angle is measured here
        c: Third point (x, y)
    
    Returns:
        Angle in degrees (0-180)
    """
    a = np.array(a, dtype=float)
    b = np.array(b, dtype=float)
    c = np.array(c, dtype=float)
    
    v1 = a - b
    v2 = c - b
    
    n1 = np.linalg.norm(v1)
    n2 = np.linalg.norm(v2)
    
    if n1 == 0 or n2 == 0:
        return 0.0
    
    cos_theta = np.dot(v1, v2) / (n1 * n2)
    cos_theta = np.clip(cos_theta, -1.0, 1.0)
    theta = np.arccos(cos_theta)
    
    return float(theta * 180.0 / np.pi)


def angle_with_vertical(p1: Tuple[float, float], p2: Tuple[float, float]) -> float:
    """
    Calculate angle between line (p1 to p2) and vertical axis.
    Used for posture analysis (shoulder-hip, hip-knee, knee-ankle with vertical).
    
    The vertical is defined as pointing downward (y increases down in image coords).
    
    Args:
        p1: Upper point (x, y) - e.g., shoulder or hip
        p2: Lower point (x, y) - e.g., hip or knee
    
    Returns:
        Angle in degrees (0 = perfectly vertical, 90 = horizontal)
    """
    p1 = np.array(p1, dtype=float)
    p2 = np.array(p2, dtype=float)
    
    # Vector from p1 to p2
    line_vec = p2 - p1
    
    # Vertical vector (pointing down in image coordinates)
    vertical = np.array([0, 1], dtype=float)
    
    line_norm = np.linalg.norm(line_vec)
    if line_norm == 0:
        return 0.0
    
    cos_theta = np.dot(line_vec, vertical) / line_norm
    cos_theta = np.clip(cos_theta, -1.0, 1.0)
    theta = np.arccos(cos_theta)
    
    return float(theta * 180.0 / np.pi)


def offset_angle(nose: Tuple[float, float], 
                 left_shoulder: Tuple[float, float], 
                 right_shoulder: Tuple[float, float]) -> float:
    """
    Calculate offset angle to detect if person is facing the camera (frontal view).
    Uses nose position relative to shoulder midpoint.
    
    Args:
        nose: Nose coordinates (x, y)
        left_shoulder: Left shoulder (x, y)
        right_shoulder: Right shoulder (x, y)
    
    Returns:
        Offset angle in degrees. Low values = side view (good for squat).
        High values = frontal view (not ideal for squat analysis).
    """
    nose = np.array(nose, dtype=float)
    left_shoulder = np.array(left_shoulder, dtype=float)
    right_shoulder = np.array(right_shoulder, dtype=float)
    
    # Midpoint of shoulders
    shoulder_mid = (left_shoulder + right_shoulder) / 2
    
    # Vector from shoulder midpoint to nose
    to_nose = nose - shoulder_mid
    
    # Shoulder line vector
    shoulder_line = right_shoulder - left_shoulder
    
    # We want to measure how much the nose is in front vs to the side
    # Use cross product to determine offset
    shoulder_width = np.linalg.norm(shoulder_line)
    if shoulder_width == 0:
        return 0.0
    
    # Calculate how far nose is from shoulder midpoint in x-direction
    # relative to shoulder width
    nose_offset = abs(to_nose[0]) / shoulder_width
    
    # Convert to angle (approximate)
    # If nose is directly above shoulder midpoint, offset is low (side view)
    # If nose is far from midpoint in x, person is at an angle
    
    # Using the approach from LearnOpenCV: angle between nose and shoulder line
    return angle_at_point(left_shoulder, nose, right_shoulder)


def get_landmark_coords(landmarks: List[dict], index: int, 
                        frame_width: float = 1.0, 
                        frame_height: float = 1.0) -> Tuple[float, float]:
    """
    Extract (x, y) coordinates from landmarks array.
    
    Args:
        landmarks: List of landmark dicts with 'x', 'y' keys (normalized 0-1)
        index: Landmark index (0-32)
        frame_width: Optional frame width for pixel conversion
        frame_height: Optional frame height for pixel conversion
    
    Returns:
        (x, y) tuple in specified coordinate space
    """
    if index >= len(landmarks):
        return (0.0, 0.0)
    
    lm = landmarks[index]
    x = lm.get('x', 0.0) * frame_width
    y = lm.get('y', 0.0) * frame_height
    
    return (x, y)


# MediaPipe landmark indices
class LandmarkIndex:
    NOSE = 0
    LEFT_SHOULDER = 11
    RIGHT_SHOULDER = 12
    LEFT_ELBOW = 13
    RIGHT_ELBOW = 14
    LEFT_WRIST = 15
    RIGHT_WRIST = 16
    LEFT_HIP = 23
    RIGHT_HIP = 24
    LEFT_KNEE = 25
    RIGHT_KNEE = 26
    LEFT_ANKLE = 27
    RIGHT_ANKLE = 28
    LEFT_FOOT = 31
    RIGHT_FOOT = 32

File: app/src/main/java/com/example/myapplication/pose/AngleUtils.kt
package com.example.myapplication.pose

import kotlin.math.acos
import kotlin.math.sqrt

/**
 * Utility functions for calculating angles from pose landmarks.
 * Ported from Python angle_utils.py
 */
object AngleUtils {
    
    /**
     * Calculate angle at point B formed by points A-B-C.
     * Used for joint angles like hip-knee-ankle (knee angle) or shoulder-elbow-wrist (elbow angle).
     *
     * @param a First point (x, y)
     * @param b Vertex point (x, y) - angle is measured here
     * @param c Third point (x, y)
     * @return Angle in degrees (0-180)
     */
    fun angleAtPoint(a: Pair<Float, Float>, b: Pair<Float, Float>, c: Pair<Float, Float>): Float {
        val v1x = a.first - b.first
        val v1y = a.second - b.second
        val v2x = c.first - b.first
        val v2y = c.second - b.second
        
        val n1 = sqrt(v1x * v1x + v1y * v1y)
        val n2 = sqrt(v2x * v2x + v2y * v2y)
        
        if (n1 == 0f || n2 == 0f) return 0f
        
        val dotProduct = v1x * v2x + v1y * v2y
        var cosTheta = dotProduct / (n1 * n2)
        cosTheta = cosTheta.coerceIn(-1f, 1f)
        
        val theta = acos(cosTheta)
        return (theta * 180f / Math.PI).toFloat()
    }
    
    /**
     * Calculate angle between line (p1 to p2) and vertical axis.
     * Used for posture analysis (shoulder-hip, hip-knee, knee-ankle with vertical).
     *
     * @param p1 Upper point (x, y) - e.g., shoulder or hip
     * @param p2 Lower point (x, y) - e.g., hip or knee
     * @return Angle in degrees (0 = perfectly vertical, 90 = horizontal)
     */
    fun angleWithVertical(p1: Pair<Float, Float>, p2: Pair<Float, Float>): Float {
        // Vector from p1 to p2
        val lineX = p2.first - p1.first
        val lineY = p2.second - p1.second
        
        // Vertical vector pointing down (y increases down in image coords)
        val verticalX = 0f
        val verticalY = 1f
        
        val lineNorm = sqrt(lineX * lineX + lineY * lineY)
        if (lineNorm == 0f) return 0f
        
        val dotProduct = lineX * verticalX + lineY * verticalY
        var cosTheta = dotProduct / lineNorm
        cosTheta = cosTheta.coerceIn(-1f, 1f)
        
        val theta = acos(cosTheta)
        return (theta * 180f / Math.PI).toFloat()
    }
    
    /**
     * Calculate offset angle to detect if person is facing the camera (frontal view).
     * Uses nose position relative to shoulder midpoint.
     *
     * @param nose Nose coordinates (x, y)
     * @param leftShoulder Left shoulder (x, y)
     * @param rightShoulder Right shoulder (x, y)
     * @return Offset angle in degrees. Low values = side view, High values = frontal view
     */
    fun offsetAngle(
        nose: Pair<Float, Float>,
        leftShoulder: Pair<Float, Float>,
        rightShoulder: Pair<Float, Float>
    ): Float {
        // Use angle at nose formed by left shoulder - nose - right shoulder
        return angleAtPoint(leftShoulder, nose, rightShoulder)
    }
    
    /**
     * Get coordinates from a landmark as a Pair.
     */
    fun getLandmarkCoords(landmarks: List<PoseLandmark>, index: Int): Pair<Float, Float> {
        if (index >= landmarks.size) return Pair(0f, 0f)
        val lm = landmarks[index]
        return Pair(lm.x, lm.y)
    }
    
    /**
     * Calculate knee angle (hip-knee-ankle)
     */
    fun calculateKneeAngle(landmarks: List<PoseLandmark>, isLeft: Boolean): Float {
        val hipIdx = if (isLeft) LandmarkIndex.LEFT_HIP else LandmarkIndex.RIGHT_HIP
        val kneeIdx = if (isLeft) LandmarkIndex.LEFT_KNEE else LandmarkIndex.RIGHT_KNEE
        val ankleIdx = if (isLeft) LandmarkIndex.LEFT_ANKLE else LandmarkIndex.RIGHT_ANKLE
        
        if (hipIdx >= landmarks.size || kneeIdx >= landmarks.size || ankleIdx >= landmarks.size) {
            return 180f
        }
        
        val hip = getLandmarkCoords(landmarks, hipIdx)
        val knee = getLandmarkCoords(landmarks, kneeIdx)
        val ankle = getLandmarkCoords(landmarks, ankleIdx)
        
        return angleAtPoint(hip, knee, ankle)
    }
    
    /**
     * Calculate elbow angle (shoulder-elbow-wrist)
     */
    fun calculateElbowAngle(landmarks: List<PoseLandmark>, isLeft: Boolean): Float {
        val shoulderIdx = if (isLeft) LandmarkIndex.LEFT_SHOULDER else LandmarkIndex.RIGHT_SHOULDER
        val elbowIdx = if (isLeft) LandmarkIndex.LEFT_ELBOW else LandmarkIndex.RIGHT_ELBOW
        val wristIdx = if (isLeft) LandmarkIndex.LEFT_WRIST else LandmarkIndex.RIGHT_WRIST
        
        if (shoulderIdx >= landmarks.size || elbowIdx >= landmarks.size || wristIdx >= landmarks.size) {
            return 180f
        }
        
        val shoulder = getLandmarkCoords(landmarks, shoulderIdx)
        val elbow = getLandmarkCoords(landmarks, elbowIdx)
        val wrist = getLandmarkCoords(landmarks, wristIdx)
        
        return angleAtPoint(shoulder, elbow, wrist)
    }
    
    /**
     * Calculate hip angle (shoulder-hip-knee)
     */
    fun calculateHipAngle(landmarks: List<PoseLandmark>, isLeft: Boolean): Float {
        val shoulderIdx = if (isLeft) LandmarkIndex.LEFT_SHOULDER else LandmarkIndex.RIGHT_SHOULDER
        val hipIdx = if (isLeft) LandmarkIndex.LEFT_HIP else LandmarkIndex.RIGHT_HIP
        val kneeIdx = if (isLeft) LandmarkIndex.LEFT_KNEE else LandmarkIndex.RIGHT_KNEE
        
        if (shoulderIdx >= landmarks.size || hipIdx >= landmarks.size || kneeIdx >= landmarks.size) {
            return 180f
        }
        
        val shoulder = getLandmarkCoords(landmarks, shoulderIdx)
        val hip = getLandmarkCoords(landmarks, hipIdx)
        val knee = getLandmarkCoords(landmarks, kneeIdx)
        
        return angleAtPoint(shoulder, hip, knee)
    }
}
